### DISCUSSION TITLE
I need to execute a file.exe with certain commands from my React Native Windows app

### DISCUSSION THREAD

**Discussion by AlmTalal** [microsoft/react-native-windows#12112] [Q&A] [answered]
Hello ! I need to execute a file.exe with certain commands. At the begining I thought that this is gonna be an easy task thanks to child-process library from node js. However this library is not available for react.

 After that i followed this [documentation](https://github.com/microsoft/react-native-winrt/blob/main/docs/USAGE.md) for being able to call this WinRT [API](https://learn.microsoft.com/en-us/uwp/api/windows.system.processlauncher?view=winrt-22621) for executing commands on the terminal. This did not work because of access denied problem (I tried by giving the windows requirment, it did not worked).
 
 My final option was to create the native module by myself, I did a module for my app (not a distributable one) in c++, this is the implmentations : 
 
```
#pragma once
#include "pch.h"
using namespace winrt;
#include <iostream>
#include <windows.h>
#include <string>
#include <sstream>
#include <stdio.h>

using namespace std;

REACT_MODULE(MyNativeModule);

struct MyNativeModule
{
    REACT_METHOD(Add, L"add");
    std::string Add(std::string path, std::string commandE) noexcept
    {
        STARTUPINFO si;
        PROCESS_INFORMATION pi;
        ZeroMemory(&si, sizeof(si));
        si.cb = sizeof(si);
        ZeroMemory(&pi, sizeof(pi));
        char cmdArgs[] = "C:\\Users\\almta\\OneDrive\\Escritorio\\IntTFHKA IntTFHKA SendCmd(!000000100000001000Item 1)";
        wchar_t wcmdArgs[256];
        MultiByteToWideChar(CP_ACP, 0, cmdArgs, -1, wcmdArgs, 256);
        if (!CreateProcess(NULL, wcmdArgs, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi)) {
            DWORD error = GetLastError();
            LPVOID errorMsg;
            FormatMessage(
                FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM,
                NULL,
                error,
                0,
                (LPWSTR)&errorMsg,
                0,
                NULL
            );
            std::wstring errorMessage(static_cast<const wchar_t*>(errorMsg));
            LocalFree(errorMsg);
            std::wcout << L"Error: " << errorMessage << std::endl;
            return "Error: " + std::to_string(error);
        }
        WaitForSingleObject(pi.hProcess, INFINITE);
        CloseHandle(pi.hProcess);
        CloseHandle(pi.hThread);
        return "success";
    }
};
```
However, here I'm having three problems : 

1. When I execute: `char cmdArgs[] = "C:\\Users\\almta\\OneDrive\\Escritorio\\IntTFHKA IntTFHKA SendCmd(!000000100000001000Item 1)";` I get `error: 5`, this means access denied.
2. When I execute: `char cmdArgs[] = "./Assets/IntTFHKA IntTFHKA SendCmd(!000000100000001000Item 1)";` I get `error : 2`, this means file not found
3. When I execute: `char cmdArgs[] ="dir";` The terminal appears, but with "_" as the only thing visible. 

Someone has any idea of how could I make this work. At this point I don't know what else to do so I'm open to any idea

**Comment by ghost** [comment]
I'm not sure if this is possible since UWP is a sandboxed environment, which restricts the usage of certain APIs, but you can check further at https://github.com/microsoft/react-native-windows/issues/11793.

**Reply by AlmTalal** [reply]
I was able to call CreateProcess, and the comand line appeared on the screen, but it still did not execute nothing

**Reply by AlmTalal** [reply]
@Victor0814gui  Is there anyway In which the app can comunicate with a file or anything that is inside the pc ?

**Reply by ghost** [reply]
I haven't tested this with WinRT C++, but the tests I conducted with C# worked well for calling certain APIs. However, I believe the project maintainers can provide a better explanation. @chrisglein

**Reply by AlmTalal** [reply]
What did you test on C#, can you share the code ?

**Reply by ghost** [reply]
My mistake, haha. I meant to talk about the projects I worked on using C# that access Win32 APIs, but I can't share the code. However, it's very similar to this one: https://github.com/microsoft/react-native-windows-samples/tree/main/samples/AppServiceDemo

**Reply by AlmTalal** [reply]
Oh thanks jajaj, I was able to access to Win32 API's however the namespace that i needed to use is forbidden

**Reply by chrisglein** [reply, marked as answer]
Short answer is that an RNW app is a UWP app and has the [same limitations](https://microsoft.github.io/react-native-windows/docs/native-code). RNW allows you to create your UI using JavaScript and React patterns, but it can only do what a UWP app can do. So the answers available about UWP apps are going to be relevant to you. From some quick searching, here are some threads:

- https://stackoverflow.com/questions/33925096/how-to-execute-process-commands-or-similar-using-a-universal-windows-platform
- https://stackoverflow.com/questions/51141474/in-which-cases-can-createprocess-be-used-on-uwp

**Comment by ivxvm** [comment]
Oh well, this is very unfortunate restriction. It seems I was wrong that I could build a beautiful and simple react-native-windows launcher for a game, time to switch to proper native languages like Rust.

**Reply by AlmTalal** [reply]
Hey, Actually they do it for the good sake of the users that wil use the developed app (executing unknown .exe files can lead to hacks). However I solved this by using react native windows WinRT modules. You can take a look at it, it may solve your problem

**Reply by stack-99** [reply]
Can you give an example of how you did it please? or which links I should look into

**Reply by AlmTalal** [reply]
You have to know what does that exe file do, after that you can use [WinRT apis](https://learn.microsoft.com/en-us/uwp/api/?view=winrt-22621) and try to copy that behavior. I don't know what exactly you want to do, but in my case it worked

**Reply by stack-99** [reply]
Well I did manage to launch a process using win32 api - CreateProcess function because I am unable to find the class FullTrustProcessLauncher in my c++ project. I tried different windows 10 sdks and it still wasn't accessible through Windows::ApplicationModel. 

Even though I managed to use CreateProcess I am still limited with the things that I can do in terms of permissions...
I can't access or create new files from external directories.

... Edit:

FullTrustProcessLauncher is the way to go (add desktop extensions as reference). Thanks!
